beacon-chain/cache/registration.go

```diff
@@ -0,0 +1,82 @@
+package cache
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+	"github.com/prysmaticlabs/prysm/v4/math"
+	ethpb "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	log "github.com/sirupsen/logrus"
+	"go.opencensus.io/trace"
+)
+
+// RegistrationCache is used to store the cached results of an Validator Registration request.
+// beacon api /eth/v1/validator/register_validator
+type RegistrationCache struct {
+	indexToRegistration map[primitives.ValidatorIndex]*ethpb.ValidatorRegistrationV1
+	lock                sync.RWMutex
+}
+
+// NewRegistrationCache initializes the map and underlying cache.
+func NewRegistrationCache() *RegistrationCache {
+	return &RegistrationCache{
+		indexToRegistration: make(map[primitives.ValidatorIndex]*ethpb.ValidatorRegistrationV1),
+		lock:                sync.RWMutex{},
+	}
+}
+
+// GetRegistrationByIndex returns the registration by index in the cache and also removes items in the cache if expired.
+func (regCache *RegistrationCache) GetRegistrationByIndex(id primitives.ValidatorIndex) (*ethpb.ValidatorRegistrationV1, error) {
```

Comment:
Optional: effective go guidelines suggest that you don't use the prefix "Get" for a getter and simply call it the thing like `regCache.RegistrationByIndex(...`

https://go.dev/doc/effective_go#Getters